File: babbleapp/src/main/java/com/learning/leap/bwb/settings/TipSettingsActivity.java
package com.learning.leap.bwb.settings;

import android.content.Intent;
import android.graphics.Bitmap;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.Switch;
import android.widget.TextView;

import com.learning.leap.bwb.BuildConfig;
import com.learning.leap.bwb.R;
import com.learning.leap.bwb.baseActivity.DetailActivity;
import com.learning.leap.bwb.baseActivity.HomeActivity;
import com.learning.leap.bwb.tipSettings.TipSettingsPresenters;
import com.learning.leap.bwb.tipSettings.TipSettingsViewInterface;
import com.learning.leap.bwb.utility.Constant;
import com.learning.leap.bwb.utility.Utility;

public class TipSettingsActivity extends AppCompatActivity implements TipSettingsViewInterface {
    TextView numberofTipsTextView;
    TextView startTimeTextView;
    TextView endTimeTextView;
    Button maxNumberOfTipsPlusButton;
    Button maxNumberOfTipsMinusButton;
    Button startTimePlusButton;
    Button startTimeMinusButton;
    Button endTimePlusButton;
    Button endTimeMinusButton;
    Button saveButton;
    Switch sendTipsTodaySwitch;
    TipSettingsPresenters tipSettingsPresenters;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fragment_tip_settings);
        setUpBackground();
        setUpViews();
        setBottomButtonViews();
        tipSettingsPresenters = new TipSettingsPresenters(this,this);
        tipSettingsPresenters.onCreate();
    }

    private void setUpViews(){
        saveButton = (Button)findViewById(R.id.tipSettingSaveButton);
        sendTipsTodaySwitch = (Switch)findViewById(R.id.tipSettingsFragmentSwitch);
        if (!BuildConfig.FLAVOR.equals("regular")) {
            sendTipsTodaySwitch.setVisibility(View.GONE);
        }
        setUpPlusAndMinusButtons();
        setUpTextViews();
        setOnClickListners();
    }

    private void setUpBackground(){
        ImageView background = (ImageView)findViewById(R.id.tipSettingsBackground);
        Bitmap backgroundBitmap = Utility.decodeSampledBitmapFromResource(getResources(),R.drawable.settings_bg,Utility.getDisplayMetrics(this));
        background.setImageBitmap(backgroundBitmap);
    }
    private void setUpPlusAndMinusButtons(){
        maxNumberOfTipsPlusButton = (Button)findViewById(R.id.tipSettingsMaxTipNumberPlusButton);
        maxNumberOfTipsMinusButton = (Button)findViewById(R.id.tipsettingsMaxNumberTipMinusButton);
        if (!BuildConfig.FLAVOR.equals("regular")) {
           maxNumberOfTipsMinusButton.setVisibility(View.GONE);
            maxNumberOfTipsPlusButton.setVisibility(View.GONE);

        }
        endTimeMinusButton = (Button)findViewById(R.id.tipSettingsEndTimeMinusButton);
        endTimePlusButton = (Button)findViewById(R.id.tipSettingsEndTimePlusButton);
        startTimeMinusButton = (Button)findViewById(R.id.tipSettingsStartTimeMinusButton);
        startTimePlusButton = (Button)findViewById(R.id.tipSettingsStartTimePlusButton);
    }
    private void setUpTextViews(){
        numberofTipsTextView = (TextView)findViewById(R.id.tipSettingsMaxNumberTipTextView);
        startTimeTextView = (TextView) findViewById(R.id.tipSettingsStartTimeTextView);
        endTimeTextView = (TextView) findViewById(R.id.tipSettingsEndTimeTextView);
        if (!BuildConfig.FLAVOR.equals("regular")) {
            numberofTipsTextView.setVisibility(View.GONE);
            TextView maxTips =  (TextView)findViewById(R.id.maxTipsTextView);
            maxTips.setVisibility(View.GONE);

        }
    }

    private void setOnClickListners(){
        saveButton.setOnClickListener(view -> tipSettingsPresenters.save());
        maxNumberOfTipsPlusButton.setOnClickListener(view -> tipSettingsPresenters.userMaxTipsPlusButtonPressed());
        maxNumberOfTipsMinusButton.setOnClickListener(view -> tipSettingsPresenters.userMaxTipMinusButtonPressed());
        startTimePlusButton.setOnClickListener(view -> tipSettingsPresenters.startTimePlusButtonPressed());
        startTimeMinusButton.setOnClickListener(view -> tipSettingsPresenters.startTimeMinusButtonPressed());
        endTimeMinusButton.setOnClickListener(view -> tipSettingsPresenters.endTimeMinusButtonPressed());
        endTimePlusButton.setOnClickListener(view -> tipSettingsPresenters.endTimePlusButtonPressed());
        sendTipsTodaySwitch.setOnCheckedChangeListener((compoundButton, b) -> tipSettingsPresenters.turnOfSwitchChangedListner(b));
    }

    private void setBottomButtonViews(){
        ImageView libraryImageView = (ImageView)findViewById(R.id.tipSettingsFragmentLibararyImageView);
        ImageView homeImageView = (ImageView)findViewById(R.id.tipSettingsFragmentHomeImageView);
        ImageView playToday = (ImageView)findViewById(R.id.tipSettingsFragmentPlayTodayImageView);
        Utility.hideButtonCheck(playToday,libraryImageView);
        libraryImageView.setOnClickListener(view -> onLibraryPress());
        homeImageView.setOnClickListener(view -> onHomePress());
        playToday.setOnClickListener(view -> onPlayToday());
    }

    @Override
    public void displayStartTimeAtIndex(String startTime) {
        startTimeTextView.setText(startTime);
    }

    @Override
    public void displayEndTimeAtIndex(String endTime) {
        endTimeTextView.setText(endTime);
    }

    @Override
    public void displayUserMaxTipsAtIndex(String maxTips) {
        numberofTipsTextView.setText(maxTips);
    }

    @Override
    public void hideStartTimePlusSigned() {
        startTimePlusButton.setEnabled(false);
    }

    @Override
    public void hideStartTimeMinusSigned() {
        startTimeMinusButton.setEnabled(false);
    }

    @Override
    public void hideEndTimePlusSigned() {
        endTimePlusButton.setEnabled(false);
    }

    @Override
    public void hideEndTimeMinusSigned() {
        endTimeMinusButton.setEnabled(false);
    }

    @Override
    public void hideMaxNumberOfTipsPlusSigned() {
        maxNumberOfTipsPlusButton.setEnabled(false);
    }

    @Override
    public void hideMaxNumberOfTipsMinusSigned() {
        maxNumberOfTipsMinusButton.setEnabled(false);
    }

    @Override
    public void setTipSwitch(Boolean sendTipsToday) {
        sendTipsTodaySwitch.setChecked(sendTipsToday);
    }

    @Override
    public void displayStartTimePlusSigned() {
        startTimePlusButton.setEnabled(true);
    }

    @Override
    public void displayStartTimeMinusSigned() {
        startTimeMinusButton.setEnabled(true);
    }

    @Override
    public void displayEndTimePlusSigned() {
        endTimePlusButton.setEnabled(true);
    }

    @Override
    public void displayEndTimeMinusSigned() {
        endTimeMinusButton.setEnabled(true);
    }

    @Override
    public void displayMaxNumberOfTipsPlusSigned() {
        maxNumberOfTipsPlusButton.setEnabled(true);
    }

    @Override
    public void displayMaxNumberOfTipsMinusSigned() {
        maxNumberOfTipsMinusButton.setEnabled(true);
    }

    @Override
    public void displaySaveButton() {
        saveButton.setVisibility(View.VISIBLE);
    }


    @Override
    public void displayErrorMessage() {
        Utility.displayAlertMessage(R.string.BabbleError,R.string.times_are_incorrect,this);
    }

    @Override
    public void saveCompleted(boolean turnOffTips) {
        if (turnOffTips){
            Utility.addCustomEvent(Constant.TURNED_OFF_NOTIFICATIONS_FOR_DAY,Utility.getUserID(this),null);
        }
        Utility.addCustomEvent(Constant.CHANGED_NOTIFICATION_SETTINGS,Utility.getUserID(this),null);
        this.finish();
    }

    @Override
    public void onHomePress() {
        Intent homeIntent = new Intent(this, HomeActivity.class);
        homeIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK| Intent.FLAG_ACTIVITY_CLEAR_TASK);
        startActivity(homeIntent);
    }

    @Override
    public void onSettingsPress() {

    }

    @Override
    public void onPlayToday() {
        Intent playTodayIntent = new Intent(this, DetailActivity.class);
        playTodayIntent.putExtra(DetailActivity.DETAIL_INTENT,DetailActivity.PLAY_TODAY);
        startActivity(playTodayIntent);
    }

    @Override
    public void onLibraryPress() {
        Intent libraryIntent = new Intent(this, DetailActivity.class);
        libraryIntent.putExtra(DetailActivity.DETAIL_INTENT,DetailActivity.LIBRARY);
        startActivity(libraryIntent);
    }
}

File: babbleapp/src/main/java/com/learning/leap/bwb/settings/TipSettingsActivity.kt
package com.learning.leap.bwb.settings

import android.content.Intent
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.learning.leap.bwb.R
import com.learning.leap.bwb.baseActivity.DetailActivity
import com.learning.leap.bwb.baseActivity.HomeActivity
import com.learning.leap.bwb.tipSettings.TipSettingsPresenters
import com.learning.leap.bwb.tipSettings.TipSettingsViewInterface
import com.learning.leap.bwb.utility.Constant
import com.learning.leap.bwb.utility.Utility
import kotlinx.android.synthetic.main.activity_tip_settings.*

class TipSettingsActivity : AppCompatActivity(), TipSettingsViewInterface {

    lateinit var tipSettingsPresenters: TipSettingsPresenters
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_tip_settings)
        supportActionBar?.setBackgroundDrawable(ColorDrawable(ContextCompat.getColor(this,R.color.dark_green)))
        supportActionBar?.title = "Tip Reminder Settings"
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        setOnClickListners()
        tipSettingsPresenters = TipSettingsPresenters(this, this)
        tipSettingsPresenters.onCreate()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId: Int = item.itemId
        if (itemId == android.R.id.home) {
            finish()
        }
        return true

    }



    private fun setOnClickListners() {
        tipSettingSaveButton.setOnClickListener { tipSettingsPresenters.save() }
        tipsettingsMaxNumberTipMinusButton.setOnClickListener { tipSettingsPresenters.userMaxTipMinusButtonPressed() }
        tipSettingsMaxTipNumberPlusButton.setOnClickListener { tipSettingsPresenters.userMaxTipsPlusButtonPressed() }
        tipSettingsStartTimeMinusButton.setOnClickListener { tipSettingsPresenters.startTimeMinusButtonPressed() }
        tipSettingsStartTimePlusButton.setOnClickListener { tipSettingsPresenters.startTimePlusButtonPressed() }
        tipSettingsEndTimeMinusButton.setOnClickListener { tipSettingsPresenters.endTimeMinusButtonPressed() }
        tipSettingsEndTimePlusButton.setOnClickListener { tipSettingsPresenters.endTimePlusButtonPressed() }
        firstTipSwitch.setOnCheckedChangeListener { _, _ ->
            tipSettingsPresenters.turnOfSwitchChangedListener()
        }
        secondTipSwitch.setOnCheckedChangeListener { _, _ ->
            tipSettingsPresenters.turnOfSwitchChangedListener()
        }

    }

    override fun displayFirstSwitch(isOn: Boolean) {
        firstTipSwitch.isChecked = isOn
    }

    override fun displaySecondSwitch(isOn: Boolean) {
        secondTipSwitch.isChecked = isOn
    }

    override fun displayStartTimeAtIndex(startTime: String) {
        tipSettingsStartTimeTextView.text = startTime
    }

    override fun displayEndTimeAtIndex(endTime: String) {
        tipSettingsEndTimeTextView.text = endTime
    }

    override fun displayUserMaxTipsAtIndex(maxTips: String) {
        tipSettingsMaxNumberTipTextView.text = maxTips
    }

    override fun hideStartTimePlusSigned() {
        tipSettingsStartTimePlusButton.isEnabled = false
    }

    override fun hideStartTimeMinusSigned() {
        tipSettingsStartTimeMinusButton.isEnabled = false
    }

    override fun hideEndTimePlusSigned() {
        tipSettingsEndTimePlusButton.isEnabled = false
    }

    override fun hideEndTimeMinusSigned() {
        tipSettingsEndTimeMinusButton.isEnabled = false
    }

    override fun hideMaxNumberOfTipsPlusSigned() {
        tipSettingsMaxTipNumberPlusButton.isEnabled = false
    }

    override fun hideMaxNumberOfTipsMinusSigned() {
        tipsettingsMaxNumberTipMinusButton.isEnabled = false
    }

    override fun setTipSwitch(sendTipsToday: Boolean) {
        //sendTipsTodaySwitch!!.isChecked = sendTipsToday
    }

    override fun displayStartTimePlusSigned() {
        tipSettingsStartTimePlusButton.isEnabled = true
    }

    override fun displayStartTimeMinusSigned() {
        tipSettingsStartTimeMinusButton.isEnabled = true
    }

    override fun displayEndTimePlusSigned() {
        tipSettingsEndTimePlusButton.isEnabled = true
    }

    override fun displayEndTimeMinusSigned() {
        tipSettingsEndTimeMinusButton.isEnabled = true
    }

    override fun displayMaxNumberOfTipsPlusSigned() {
        tipSettingsMaxTipNumberPlusButton.isEnabled = true
    }

    override fun displayMaxNumberOfTipsMinusSigned() {
        tipsettingsMaxNumberTipMinusButton.isEnabled = true
    }

    override fun displaySaveButton() {
        tipSettingSaveButton.visibility = View.VISIBLE
    }

    override fun displayErrorMessage() {
        Utility.displayAlertMessage(R.string.BabbleError, R.string.times_are_incorrect, this)
    }

    override fun tipOneSwitchIsOn(): Boolean {
        return firstTipSwitch.isChecked
    }

    override fun tipTwoSwitchIsOn(): Boolean {
        return secondTipSwitch.isChecked
    }

    override fun saveCompleted(turnOffTips: Boolean) {
        if (turnOffTips) {
            Utility.addCustomEvent(Constant.TURNED_OFF_NOTIFICATIONS_FOR_DAY, Utility.getUserID(this), null)
        }
        Utility.addCustomEvent(Constant.CHANGED_NOTIFICATION_SETTINGS, Utility.getUserID(this), null)
        finish()
    }

    override fun onHomePress() {
        val homeIntent = Intent(this, HomeActivity::class.java)
        homeIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
        startActivity(homeIntent)
    }

    override fun onSettingsPress() {}
    override fun onPlayToday() {
        val playTodayIntent = Intent(this, DetailActivity::class.java)
        playTodayIntent.putExtra(DetailActivity.DETAIL_INTENT, DetailActivity.PLAY_TODAY)
        startActivity(playTodayIntent)
    }

    override fun onLibraryPress() {
        val libraryIntent = Intent(this, DetailActivity::class.java)
        libraryIntent.putExtra(DetailActivity.DETAIL_INTENT, DetailActivity.LIBRARY)
        startActivity(libraryIntent)
    }
}